modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxUnlockRequestSender.java

```diff
@@ -0,0 +1,171 @@
+import org.apache.ignite.internal.tx.impl.TxManagerImpl.TransactionFailureHandler;
+import org.apache.ignite.internal.tx.message.LockReleaseMessage;
+import org.apache.ignite.internal.tx.message.TxMessagesFactory;
+import org.apache.ignite.internal.util.CompletableFutures;
+import org.apache.ignite.network.ClusterService;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Sends TX Unlock request.
+ */
+public class TxUnlockRequestSender {
+
+    /** Network timeout. */
+    private static final long RPC_TIMEOUT = 3000;
+
+    /** Tx messages factory. */
+    private static final TxMessagesFactory FACTORY = new TxMessagesFactory();
+
+    /** Cluster service. */
+    private final ClusterService clusterService;
+
+    /** Placement driver helper. */
+    private final PlacementDriverHelper placementDriverHelper;
+
+    /** Hybrid clock. */
+    private final HybridClock hybridClock;
+
+    /** Cleanup processor. */
+    private final TxCleanupProcessor txCleanupProcessor;
+
+    /**
+     * The constructor.
+     *
+     * @param clusterService Cluster service.
+     * @param placementDriverHelper Placement driver helper.
+     * @param clock A hybrid logical clock.
+     * @param txCleanupProcessor A cleanup processor.
+     */
+    public TxUnlockRequestSender(
+            ClusterService clusterService,
+            PlacementDriverHelper placementDriverHelper,
+            HybridClock clock,
+            TxCleanupProcessor txCleanupProcessor) {
+        this.clusterService = clusterService;
+        this.placementDriverHelper = placementDriverHelper;
+        this.hybridClock = clock;
+        this.txCleanupProcessor = txCleanupProcessor;
+    }
+
+    /**
+     * Sends unlock request to the primary nodes of each one of {@code partitions}.
+     *
+     * @param partitions Enlisted partition groups.
+     * @param commit {@code true} if a commit requested.
+     * @param commitTimestamp Commit timestamp ({@code null} if it's an abort).
+     * @param txId Transaction id.
+     * @return Completable future of Void.
+     */
+    public CompletableFuture<Void> unlock(
+            Collection<TablePartitionId> partitions,
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp,
+            UUID txId
+    ) {
+        return placementDriverHelper.findPrimaryReplicas(partitions, hybridClock.now())
+                .thenCompose(partitionData -> {
+                    durableCleanupPartitionsWithNoPrimary(commit, commitTimestamp, txId, partitionData.partitionsWithoutPrimary);
+
+                    return unlockPartitions(partitionData.partitionsByNode, commit, commitTimestamp, txId);
+                });
+    }
+
+    private void durableCleanupPartitionsWithNoPrimary(
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp,
+            UUID txId,
+            Set<TablePartitionId> noPrimaryFound
+    ) {
+        if (!noPrimaryFound.isEmpty()) {
```

Comment:
Why we need it? Seems that iteration itself over an empty set will do the trick, right?

Reply:
> no concerns regarding excess memory usage.
Curios whether it really matters. Is it a best pattern by default or there was a benchmark?